Tests/SUSignatureVerifierTest.m

```diff
@@ -74,9 +105,89 @@ - (BOOL)checkFile:(NSString *)aFile withPubKey:(NSString *)pubKey signature:(NSS
     return [v verifyFileAtPath:aFile signatures:sig];
 }
 
+- (BOOL)checkFile:(NSString *)aFile withEdKey:(NSString *)pubKey signature:(NSString *)sigString
+{
+    SUPublicKeys *pubKeys = [[SUPublicKeys alloc] initWithDsa:nil ed:pubKey];
+    SUSignatureVerifier *v = [[SUSignatureVerifier alloc] initWithPublicKeys:pubKeys];
+
+    SUSignatures *sig = [[SUSignatures alloc] initWithDsa:nil ed:sigString];
+
+    return [v verifyFileAtPath:aFile signatures:sig];
+}
+
+- (void)testVerifyFileWithBothKeys
+{
+    NSString *dsaKey = [NSString stringWithContentsOfFile:self.pubDSAKeyFile encoding:NSASCIIStringEncoding error:nil];
+    XCTAssertNotNil(dsaKey, @"Public key must be readable");
+
+    SUPublicKeys *pubKeys = [[SUPublicKeys alloc] initWithDsa:dsaKey ed:self.pubEdKey];
+    SUSignatureVerifier *v = [[SUSignatureVerifier alloc] initWithPublicKeys:pubKeys];
+
+    XCTAssertFalse([v verifyFileAtPath:self.testFile
+                            signatures:[[SUSignatures alloc] initWithDsa:nil ed:nil]],
+                   @"Fail if no signatures are provided");
+    XCTAssertFalse([v verifyFileAtPath:self.testFile
+                            signatures:[[SUSignatures alloc] initWithDsa:@"lol" ed:@"lol"]],
+                   @"Fail if both signatures are invalid");
+
+    NSString *dsaSig = @"MCwCFCIHCIYYkfZavNzTitTW5tlRp/k5AhQ40poFytqcVhIYdCxQznaXeJPJDQ==";
+    NSString *wrongDSASig = @"MCwCFCIHCiYYkfZavNzTitTW5tlRp/k5AhQ40poFytqcVhIYdCxQznaXeJPJDQ==";
+    NSString *edSig = @"EIawm2YkDZ2gBfkEMF2+1VuuTeXnCGZOdnMdVgPPvDZioq7bvDayXqKkIIzSjKMmeFdcFJOHdnba5ZV60+gPBw==";
+    NSString *wrongEdSig = @"wTcpXCgWoa4NrJpsfzS61FXJIbv963//12U2ef9xstzVOLPHYK2N4/ojgpDV5N1/NGG1uWMBgK+kEWp0Z5zMDQ==";
+
+    XCTAssertTrue([v verifyFileAtPath:self.testFile
+                           signatures:[[SUSignatures alloc] initWithDsa:dsaSig ed:nil]],
+                  @"Allow just a DSA signature if that's all that's available");
+    XCTAssertFalse([v verifyFileAtPath:self.testFile
+                            signatures:[[SUSignatures alloc] initWithDsa:nil ed:edSig]],
+                   @"Require the DSA signature to match because there's a DSA public key");
+
+    XCTAssertFalse([v verifyFileAtPath:self.testFile
+                            signatures:[[SUSignatures alloc] initWithDsa:dsaSig ed:wrongEdSig]],
+                   @"Fail on a bad Ed25519 signature regardless");
+    XCTAssertFalse([v verifyFileAtPath:self.testFile
+                            signatures:[[SUSignatures alloc] initWithDsa:wrongDSASig ed:edSig]],
+                   @"Fail on a bad DSA signature if provided");
+
+    XCTAssertFalse([v verifyFileAtPath:self.testFile
+                            signatures:[[SUSignatures alloc] initWithDsa:dsaSig ed:@"lol"]],
+                   @"Fail if the Ed25519 signature is invalid.");
```

Comment:
It's this test that used to fail: verification would *succeed* if both signatures were present but the Ed25519 one was invalid. I consider this a bug because it means someone could think they were using Ed25519 in their appcast, but if they had a typo Sparkle would silently fall back to only checking the DSA signature. (I believe this would be a bug in 1.x as well.)

Perhaps I should have titled the PR around this bug rather than around adding tests; it's the only bug I discovered while adding these tests.

Reply:
Maybe it should SULog when an invalid signature is decoded?

Reply:
Failing on syntactically-invalid keys is a good behavior.

Logging invalid keys would be great.